HIFIMAGS/HIFIMAGS-IOC-01App/src/hifimagsys.st

```diff
@@ -0,0 +1,266 @@
+program hifimagsys("P")
+
+#include "seqPVmacros.h"
+%% #include "errlog.h"
+
+/* Turn on run-time debug messages */
+option +d;
+
+/* Make code reentrant. This is needed to run more than one instance of this program. */
+option +r;
+
+/* Variables */
+int prev_opmode = -1;
+int i;
+int length;
+string pvName;
+int prevdis;
+
+int setmag;
+double target;
+int disable;
+float val;
+assign setmag to "";
+assign target to "";
+assign disable to "";
+assign val to "";
+
+char mags[4] = {'X', 'Y', 'Z', 'M'};
+char inactive_in_high[2] = {'X', 'Y'};
+char zero_in_high[3] = {'X', 'Y', 'Z'};
+char set_zerofield[3] = {'X', 'Y', 'Z'};
+char zero_in_switching[2] = {'X', 'Y'};
+char inactive_in_switching[3] = {'X', 'Y', 'Z'};
+
+/* Environment Variables */
+string *prefix;
+
+/* PVs */
+PV(int, magoff, "{P}MAGNETS:OFF:SP", Monitor);
+PV(int, opmode, "{P}OPMODE", Monitor);
+PV(int, setopmode, "{P}OPMODE:SP", NoMon);
+PV(int, xquench, "{P}X:QUENCH", Monitor);
+PV(int, yquench, "{P}Y:QUENCH", Monitor);
+PV(int, zquench, "{P}Z:QUENCH", Monitor);
+PV(int, mquench, "{P}M:QUENCH", Monitor);
+PV(int, mextradis, "{P}M:EXTRAS:DIS", NoMon);
+PV(int, zswitchdis, "{P}Z:SWITCH:DIS", NoMon);
+PV(int, mramp, "{P}M:RAMP:LEADS:SP", NoMon);
+PV(int, mpersist, "{P}M:PERSIST:SP", NoMon);
+
+/* STATE SETS */
+
+/* Magnets Off - Ramp all magnets to 0 */
+ss systemManager {
+
+  state init {
+    when () {
+      prefix = macValueGet("P");
+    } state monitoring
+  }
+
+  state monitoring {
+    when (xquench == 1 || yquench == 1 || zquench == 1 || mquench == 1){
+    } state quench
+    when (magoff == 1) {
+    } state turnmagoff
+    when (opmode != prev_opmode) {
+      prev_opmode = opmode;
+    } state opmode
+    when (opmode == prev_opmode) {      
+    } state monitoring
```

Comment:
I think this is probably what causes a busy-loop as it will keep re-entering this state with no delays. Do we actually need this or can it just wait until one of the other conditions is triggered?

Reply:
Without this, there is something which impacts the behaviour of the operation modes that I can't make sense of. I have just double-checked and this seems to be the only solution that seems to work with the tests - if you have other ideas I'm willing to listen.

Reply:
ok - I will look into this. Initial impression is that it's probably a race condition somewhere as the tests fails inconsistently with it removed.